File: apps/block-proxy/src/upstream/near_lake.rs
use std::time::Duration;

use aws_sdk_s3::config::Region;
use bytes::Bytes;
use futures::future::try_join_all;
use tracing::debug;

use crate::config::Config;

/// NEAR Lake upstream fetcher with shard assembly.
///
/// NEAR Lake stores blocks as multiple separate S3 objects:
/// - `{padded_height}/block.json` — block header and metadata
/// - `{padded_height}/shard_{i}.json` — one object per shard
///
/// This fetcher retrieves all objects concurrently and assembles them into the
/// fastnear-compatible single-object format: `{"block": ..., "shards": [...]}`.
/// This enables the fallback chain to treat all upstream sources uniformly (PRXY-09).
pub struct NearLakeUpstream {
    client: aws_sdk_s3::Client,
    bucket: String,
    timeout: Duration,
}

impl NearLakeUpstream {
    /// Creates a new `NearLakeUpstream` from configuration.
    ///
    /// Uses the default AWS credential chain (env vars, instance metadata, etc.) —
    /// NOT MinIO credentials. NEAR Lake data is hosted on real AWS S3.
    ///
    /// Region defaults to "eu-central-1" (where NEAR Lake S3 buckets are hosted).
    pub async fn new(config: &Config) -> Self {
        let region = Region::new(config.near_lake_region.clone());

        let sdk_config = aws_config::defaults(aws_config::BehaviorVersion::latest())
            .region(region)
            .load()
            .await;

        let client = aws_sdk_s3::Client::new(&sdk_config);

        Self {
            client,
            bucket: config.near_lake_bucket_name(),
            timeout: Duration::from_secs(config.upstream_timeout_secs),
        }
    }

    /// Fetches a single S3 object and returns its body bytes.
    ///
    /// Returns a descriptive error if the key does not exist (NoSuchKey).
    /// Takes an owned String so the future does not borrow from the caller.
    async fn get_object_owned(&self, key: String) -> anyhow::Result<Bytes> {
        let output = self
            .client
            .get_object()
            .bucket(&self.bucket)
            .key(&key)
            .request_payer(aws_sdk_s3::types::RequestPayer::Requester)
            .send()
            .await
            .map_err(|err| {
                if let Some(service_err) = err.as_service_error() {
                    if service_err.is_no_such_key() {
                        return anyhow::anyhow!("NEAR Lake object not found: {}", key);
                    }
                }
                anyhow::anyhow!("NEAR Lake get_object failed for key {}: {}", key, err)
            })?;

        let data = output
            .body
            .collect()
            .await
            .map_err(|e| anyhow::anyhow!("NEAR Lake body collect failed for key {}: {}", key, e))?
            .into_bytes();

        if data.len() > 100 * 1024 * 1024 {
            anyhow::bail!("NEAR Lake response too large for key {}: {} bytes", key, data.len());
        }

        Ok(data)
    }

    /// Fetches a block by height from NEAR Lake and assembles the shard data.
    ///
    /// Steps:
    /// 1. Zero-pad height to 12 digits
    /// 2. Fetch `{padded}/block.json`
    /// 3. Read actual shard IDs from `block.chunks[].shard_id` (matching
    ///    the approach used by near-lake-framework-rs — shard IDs may not
    ///    be sequential after dynamic resharding)
    /// 4. Fetch all `{padded}/shard_{id}.json` concurrently via try_join_all
    /// 5. Assemble into `{"block": <block>, "shards": [<shard_0>, ...]}`
    /// 6. Serialize and return as Bytes
    ///
    /// The entire operation is wrapped in `tokio::time::timeout` to enforce the
    /// per-source timeout (PRXY-07).
    pub async fn fetch(&self, height: u64) -> anyhow::Result<Bytes> {
        let start = std::time::Instant::now();

        let result = tokio::time::timeout(self.timeout, async {
            let padded = format!("{:012}", height);

            // Fetch block.json
            let block_key = format!("{}/block.json", padded);
            let block_bytes = self.get_object_owned(block_key).await?;

            // Parse block to extract shard IDs from chunks array.
            let block_value: serde_json::Value = serde_json::from_slice(&block_bytes)
                .map_err(|e| anyhow::anyhow!("NEAR Lake block.json parse failed for height {}: {}", height, e))?;

            // Extract actual shard IDs from block.chunks[].shard_id.
            // This matches how near-lake-framework-rs determines which shard
            // files to fetch — shard IDs may not be sequential (e.g. after
            // dynamic resharding).
            let chunks = block_value
                .get("chunks")
                .and_then(|c| c.as_array())
                .ok_or_else(|| anyhow::anyhow!(
                    "NEAR Lake block.json missing chunks array for height {}", height
                ))?;

            let shard_ids: Vec<u64> = chunks
                .iter()
                .enumerate()
                .map(|(i, chunk)| {
                    chunk.get("shard_id")
                        .and_then(|v| v.as_u64())
                        .ok_or_else(|| anyhow::anyhow!(
                            "NEAR Lake chunk {} missing shard_id for height {}", i, height
                        ))
                })
                .collect::<anyhow::Result<Vec<_>>>()?;

            // Fetch all shards concurrently using actual shard IDs
            let shard_futures: Vec<_> = shard_ids
                .iter()
                .map(|&id| {
                    let key = format!("{}/shard_{}.json", padded, id);
                    self.get_object_owned(key)
                })
                .collect();

            let shard_bytes_vec: Vec<Bytes> = try_join_all(shard_futures)
                .await
                .map_err(|e| anyhow::anyhow!("NEAR Lake shard fetch failed for height {}: {}", height, e))?;

            // Parse each shard as raw JSON
            let shards: Vec<serde_json::Value> = shard_bytes_vec
                .iter()
                .enumerate()
                .map(|(i, shard_bytes)| {
                    serde_json::from_slice(shard_bytes).map_err(|e| {
                        anyhow::anyhow!("NEAR Lake shard_{}.json parse failed for height {}: {}", shard_ids[i], height, e)
                    })
                })
                .collect::<anyhow::Result<Vec<_>>>()?;

            // Assemble into fastnear-compatible shape: {"block": ..., "shards": [...]}
            let assembled = serde_json::json!({
                "block": block_value,
                "shards": shards,
            });

            let serialized = serde_json::to_string(&assembled)
                .map_err(|e| anyhow::anyhow!("NEAR Lake assembly serialization failed for height {}: {}", height, e))?;

            Ok::<Bytes, anyhow::Error>(Bytes::from(serialized))
        })
        .await
        .map_err(|_| anyhow::anyhow!("NEAR Lake fetch timed out for block {} after {:?}", height, self.timeout))??;

        let elapsed = start.elapsed();
        debug!(
            height,
            bytes = result.len(),
            latency_ms = elapsed.as_millis(),
            "NEAR Lake upstream fetch complete"
        );

        Ok(result)
    }
}

File: apps/block-proxy/src/upstream/near-lake.ts
import { GetObjectCommand, S3Client } from '@aws-sdk/client-s3';

import { logger } from 'nb-logger';

import type { Config } from '#config';

const NEAR_LAKE_LAG_DEPTH = 10;
const POLL_INTERVAL_MS = 1000;

export class NearLakeUpstream {
  private bucket: string;
  private client: S3Client;
  private timeoutMs: number;

  private constructor(config: Config) {
    this.bucket = config.nearLakeBucket;
    this.timeoutMs = config.upstreamTimeoutMs;

    this.client = new S3Client({
      credentials: {
        accessKeyId: config.nearLakeAccessKey,
        secretAccessKey: config.nearLakeSecretKey,
      },
      region: config.nearLakeRegion,
    });
  }

  static create(config: Config): NearLakeUpstream | null {
    if (!config.nearLakeAccessKey || !config.nearLakeSecretKey) {
      return null;
    }
    return new NearLakeUpstream(config);
  }

  private async fetchAssembled(
    height: number,
    abortSignal?: AbortSignal,
  ): Promise<Buffer> {
    const padded = String(height).padStart(12, '0');

    // Fetch block.json
    const blockKey = `${padded}/block.json`;
    const blockBytes = await this.getObject(blockKey, abortSignal);
    const blockValue = JSON.parse(blockBytes.toString('utf8'));

    // Extract actual shard IDs from block.chunks[].shard_id
    const chunks = blockValue?.chunks;
    if (!Array.isArray(chunks)) {
      throw new Error(
        `NEAR Lake block.json missing chunks array for height ${height}`,
      );
    }

    const shardIds: number[] = chunks.map(
      (chunk: { shard_id?: number }, i: number) => {
        if (typeof chunk.shard_id !== 'number') {
          throw new Error(
            `NEAR Lake chunk ${i} missing shard_id for height ${height}`,
          );
        }
        return chunk.shard_id;
      },
    );

    // Fetch all shards concurrently
    const shardBuffers = await Promise.all(
      shardIds.map((id) =>
        this.getObject(`${padded}/shard_${id}.json`, abortSignal),
      ),
    );

    const shards = shardBuffers.map((buf, i) => {
      try {
        return JSON.parse(buf.toString('utf8'));
      } catch (err) {
        throw new Error(
          `NEAR Lake shard_${shardIds[i]}.json parse failed for height ${height}: ${err}`,
        );
      }
    });

    // Assemble into fastnear-compatible shape: {"block": ..., "shards": [...]}
    const assembled = { block: blockValue, shards };
    return Buffer.from(JSON.stringify(assembled));
  }

  private async getObject(
    key: string,
    abortSignal?: AbortSignal,
  ): Promise<Buffer> {
    let output;
    try {
      output = await this.client.send(
        new GetObjectCommand({
          Bucket: this.bucket,
          Key: key,
          RequestPayer: 'requester',
        }),
        abortSignal ? { abortSignal } : undefined,
      );
    } catch (err: unknown) {
      const name = (err as { name?: string })?.name;
      if (name === 'NoSuchKey') {
        const notFoundErr = new Error(
          `NEAR Lake object not found: ${key}`,
        ) as Error & { notFound: boolean };
        notFoundErr.notFound = true;
        throw notFoundErr;
      }
      throw new Error(`NEAR Lake get_object failed for key ${key}: ${err}`);
    }

    if (!output.Body) {
      throw new Error(`NEAR Lake empty body for key ${key}`);
    }

    const bytes = await output.Body.transformToByteArray();
    if (bytes.length > 100 * 1024 * 1024) {
      throw new Error(
        `NEAR Lake response too large for key ${key}: ${bytes.length} bytes`,
      );
    }

    return Buffer.from(bytes);
  }

  async fetch(
    height: number,
    getTipHeight: () => number = () => 0,
  ): Promise<Buffer> {
    const start = Date.now();
    const abort = new AbortController();
    const timer = setTimeout(() => abort.abort(), this.timeoutMs);

    try {
      try {
        const result = await this.fetchAssembled(height, abort.signal);
        logger.debug(
          { bytes: result.length, height, latency_ms: Date.now() - start },
          'NEAR Lake upstream fetch complete',
        );
        return result;
      } catch (firstErr) {
        if (!(firstErr as Error & { notFound?: boolean }).notFound) {
          throw firstErr;
        }
      }

      const tipHeight = getTipHeight();

      // Height too far ahead of known tip
      if (tipHeight > 0 && height > tipHeight + NEAR_LAKE_LAG_DEPTH) {
        const err = new Error(
          `NEAR Lake block ${height} is too far ahead of tip ${tipHeight}`,
        ) as Error & { notFound: boolean };
        err.notFound = true;
        throw err;
      }

      // Height at or below known tip — block is finalized but absent from S3 (skipped)
      if (tipHeight > 0 && height <= tipHeight) {
        logger.debug(
          { height, tipHeight },
          'NEAR Lake block missing below tip, treating as skipped block',
        );
        return Buffer.from('null');
      }

      // Height near-future block — wait proportional to gap, then retry once
      const gapSeconds =
        tipHeight > 0
          ? Math.min(height - tipHeight, NEAR_LAKE_LAG_DEPTH)
          : NEAR_LAKE_LAG_DEPTH;
      const waitMs = gapSeconds * POLL_INTERVAL_MS;

      logger.debug(
        {
          gapSeconds,
          height,
          latency_ms: Date.now() - start,
          tipHeight,
          waitMs,
        },
        'NEAR Lake block ahead of tip, waiting before retry',
      );

      await Promise.race([
        new Promise<void>((resolve) => setTimeout(resolve, waitMs)),
        new Promise<void>((resolve) =>
          abort.signal.addEventListener('abort', () => resolve(), {
            once: true,
          }),
        ),
      ]);

      if (abort.signal.aborted) {
        throw new Error(
          `NEAR Lake fetch timed out for block ${height} after ${this.timeoutMs}ms`,
        );
      }

      try {
        const result = await this.fetchAssembled(height, abort.signal);
        logger.debug(
          { bytes: result.length, height, latency_ms: Date.now() - start },
          'NEAR Lake upstream fetch complete on retry',
        );
        return result;
      } catch (retryErr) {
        if (!(retryErr as Error & { notFound?: boolean }).notFound) {
          throw retryErr;
        }
        logger.debug(
          { gapSeconds, height, tipHeight },
          'NEAR Lake block not found after wait, treating as skipped block',
        );
        return Buffer.from('null');
      }
    } finally {
      clearTimeout(timer);
    }
  }
}
